File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        int nn = n;
        if (n < 0) {
            nn = n * -1;
        }
        double ans = 1.0;
        while (nn >= 1) {
            if (nn % 2 == 0) {
                x = x * x;
                nn = nn / 2;
            } else {
                ans = ans * x;
                x = x * x;
                nn = nn / 2;
            }
        }
        if (n < 0) {
            return 1.0 / ans;
        }
        return ans;
    }
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        sign=-1 if n<0 else 1
        n=n*-1 if n<0 else n
        power=self.power(x,n)
        return power if sign==1 else 1/power
    def power(self,x,n):
        if n==0:
            return 1
        pow=self.power(x,n//2)
        if n&1==0:
            return pow*pow
        else:
            return x*pow*pow


        